sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java

```diff
@@ -341,7 +348,12 @@ public PipelineResult run(PipelineOptions options) {
     final PipelineResult pipelineResult;
     try {
       enforcement.get().beforePipelineExecution();
-      pipelineResult = super.run(options);
+      PipelineOptions updatedOptions =
+          MAPPER.convertValue(MAPPER.valueToTree(options), PipelineOptions.class);
+      updatedOptions
+          .as(TestValueProviderOptions.class)
+          .setProviderRuntimeValues(StaticValueProvider.of(providerRuntimeValues));
```

Comment:
Does this really work? One of the ways pipeline options are used is that the user captures and serializes the RuntimeValueProvider, and then its value is changed at runtime. This looks like you are changing the *instance* of the ValueProvider associated with an option, rather than the value the instance returns.

Reply:
Curiously enough (or perhaps morbidly enough) this is exactly what happens when you run a template in Dataflow: we update the JSON pipeline options with concrete values, and they deserialize as StaticValueProvider's (via ValueProvider.Deserializer). And then the serialized RuntimeValueProvider.get() detects that it's a StaticValueProvider and returns its value.